### PR TITLE
Pin navigation2 repos and simplify docker build (#142).

### PR DESCRIPTION
**Pull request by xfiderek**
This PR addresses: https://github.com/space-ros/docker/issues/142

This PR introduces the following changes
- Pin version of navigation2.repos to ensure reproducible builds
- Create update script according to diagram from here: https://github.com/space-ros/space-ros/discussions/166#discussioncomment-9894710. The script is called `docker_update_nav2_repos.sh`
- Simplify dockerfile
- Add `space_nav2_bringup` package allowing to launch nav2 stack (we had to get rid of nav2_bringup due to heavy dependencies such as rviz)
- Update nav2_demo accordingly -> update readme

Out of scope (will be fixed as part of different issues):
- Simplifying space_robots demo (e.g. excluding gazebo from space robots img)
- Optimizing image size of nav2 image -> will be addressed separately
- Move the docker image to stacks repo -> will be addressed separately
- Remove rviz2 from the image

### CODE REVIEW

**Comment by xfiderek** [comment]
Hi @mkhansenbot, thanks for the review!

Indeed, you are right that there are lots of changes in this PR and I am fine with splitting it further. However, doing as you suggested, I believe the demo will stop working after merging PR#2 and before PR#3 is merged, since rviz will no longer be available in the nav2 image (unless we want to launch it in a separate container, however this would require some changes to other dockerfiles that would later be reverted). Am I right?
If that is the case, I propose to split this work into 2 PRs instead of 3:
* PR#1: pin the navigation2.repos file, add the scripts to generate it, and update the Dockerfile (closing #142)
* PR#2: Remove rviz dependency, add space_nav2_bringup, update demo to use separate image from rviz (closing https://github.com/space-ros/demos/issues/28)

What do you think?

**Comment by xfiderek** [comment]
hi @Bckempa. I plan to finish it this weekend, but would be good to merge https://github.com/space-ros/space-ros/pull/170 because b4 that we can't move it from draft stage

**Comment by asimonov** [comment]
ah, I never looked at this PR in detail, but looks like I was working on similar issues with #165 to cleanup some of dependencies management.

happy to revert PR I created, but I am curious what is the plan for this PR? split?

File: navigation2/generate_repos.sh
Comment (asimonov) [inline navigation2/generate_repos.sh:1]: is this a copy of this script from space-ros? i think they are slighly different... not sure if it is great to keep two copies in two repos, but not sure how to best resolve...

File: navigation2/generate_repos.sh [hunk 1]
```diff
@@ -0,0 +1,142 @@
+#!/bin/bash
+
+# TODO - remove this file from here as soon as https://github.com/eholum/space-ros/pull/1 is merged
```
Comment (asimonov) [inline navigation2/generate_repos.sh:3]: ah, i see, it will be gone...

**Comment by asimonov** [comment]
May be more general question is in order here... 

we are using rosinstall_generator (not just here, but in space-ros too), but the ros2 core has not fully switched to it yet, but uses ros2.repos file (https://docs.ros.org/en/jazzy/Installation/Alternatives/Ubuntu-Development-Setup.html#build-ros-2).

The issue was discussed here: https://github.com/ros2/ros2_documentation/issues/1470, but it was de-scoped from humble.

Does anyone know the status of this? rosinstall_generator obviously works for ros2, but it is still not official way...

File: navigation2/README.md [hunk 2]
```diff
@@ -37,13 +38,28 @@ There is a run.sh script provided for convenience that will run the spaceros ima
...
+To run the latest demo, see the README in the [nav2_demo](../nav2_demo/README.md) folder.
+
+
+## Updating navigation2 packages
+The `navigation2.repos` file available in this repository provides a list of repos that are required to build the nav2 stack. ROS packages already included in the SpaceROS base image are omited from this file.
+
+To update navigation2 packages used to build the space nav2 image, run the `docker_update_nav2_repos.sh` script:
+
+```
+./docker_update_nav2_repos.sh
```
Comment (asimonov) [inline navigation2/README.md:59]: I am trying to run this script. i have a local osrf/space-ros:latest that I just built from space-ros main branch (including PR 170).

but i get an error:

```
~/git/space-ros/docker/navigation2 (pr152_pin_nav2)$ ./docker_update_nav2_repos.sh 
Error: Either packages or repos (or both) have to be specified
Generate .repos file with list of SpaceROS repositories to clone
Usage: ./generate_repos.sh --packages PACKAGES --outfile OUTFILE [--excluded-packages EXCLUDED_PACKAGES] [--repos REPOS] [--rosdistro ROSDISTRO] [--upstream UPSTREAM] [--exclude-installed EXCLUDE_INSTALLED]
  --outfile                   The output file
  --packages                  File containing a list of ROS packages to include
  --repos                     File containing a list of ROS repos to use.
                              Use as alternative to 'packages', rosinstall can generate .repos file from list of repos instead of list of packages
  --excluded-packages         File containing a list of packages to exclude [optional] 
  --rosdistro                 ROS2 distribution (default: humble)
  --upstream                  Specify whether to use version tags of upstream repositories (default: true)
  --exclude-installed         Whether to exclude already installed packages from .repos file. Installed workspaces must be sourced to make it work (default: true)
  -h, --help                  Display this help and exit
```

Comment (asimonov) [reply navigation2/README.md:59]: ah, $ROS_DISTRO vs $ROSDISTRO in `update_nav2_repos.sh`

**Comment by xfiderek** [comment]
> ah, I never looked at this PR in detail, but looks like I was working on similar issues with #165 to cleanup some of dependencies management.
> 
> happy to revert PR I created, but I am curious what is the plan for this PR? split?

Yes, I will split the PR into 2 according to what I discussed with @mkhansenbot above.

File: navigation2/README.md (hunk 2 above)
Comment (xfiderek) [reply navigation2/README.md:59]: Ok, so it seems that I need to fix it. Will run it again and check if it works. I plan to solve it today/tomorrow

**Comment by asimonov** [comment]
> > ah, I never looked at this PR in detail, but looks like I was working on similar issues with #165 to cleanup some of dependencies management.
> > happy to revert PR I created, but I am curious what is the plan for this PR? split?
> 
> Yes, I will split the PR into 2 according to what I discussed with @mkhansenbot above + gazebo/rviz will be removed from nav2 img

ok, I would like to test it, but will wait till you split.

**Comment by xfiderek** [comment]
> > 
> 
> @xfiderek - if you remove Gazebo from the Nav2 build, the system tests (which actually run simulations) will fail. I thought the plan was to remove rviz. Can we do that and leave Gazebo alone for now? I don't want a regression to occur. At a later date, we can add our own system tests and remove Gazebo

@mkhansenbot my bad, I don't plan on removing gazebo here. In this PR I am removing only rviz dependency. Gazebo stays as-is for now. I am working on splitting this PR right now.

So overall:
- This PR will only pin nav2 repos and address the corresponding issue
- A separate PR will be created by me remove Rviz as dependency. We need to file a new issue for that or use #165 
- Yet another issue should be created to remove gazebo as a dependency and we can work on it separately.

**Comment by xfiderek** [comment]
Hi @mkhansenbot , @Bckempa, @asimonov, other potential reviewers

I tried dividing it as we initially anticipated, but due to the problem with apt/from-source installation (dependency on nav2_bringup) it is really hard to make pinned repos working accurately without introducing space_nav2_bringup package. Is it ok if I introduce both pinned repos and space_nav2_bringup as part of this PR?  Otherwise I would need to rework this significantly, only to rollback to what we have here. I will keep rviz2 in this PR to keep reworks to the demo minimal.

I tested it locally and it is ready for review, but I think we need to wait for https://github.com/space-ros/docker/pull/162 before merging it, as we can't test it with the main version of spaceros image from main now due to using $ROSDISTRO instead of $ROS_DISTRO there.

So to sum it up:
- this PR is ready for review - it introduces space_nav2_bringup and pin repos, closing #142 
- my next PR (which I will create after this one gets merged) will remove dep on rviz2 and close https://github.com/space-ros/demos/issues/28
- one more PR and corresponding issue will be created to remove gazebo from space_robots

File: navigation2/README.md (hunk 2 above)
Comment (xfiderek) [reply navigation2/README.md:59]: Fixed

File: navigation2/Dockerfile [hunk 3]
```diff
@@ -40,67 +40,33 @@ ARG DEBIAN_FRONTEND=noninteractive
...
-
-# Build Navigation2
-SHELL ["/bin/bash", "-c"]
-RUN source ${NAV2_DEPS_WS}/install/setup.bash && colcon build --cmake-args -DCMAKE_BUILD_TYPE=Release -DCMAKE_EXPORT_COMPILE_COMMANDS=ON
-
-# Install rviz2 to send goals to Nav2
-RUN sudo apt update && sudo apt install -y ros-humble-rviz2
+COPY navigation2.repos .
+COPY excluded-pkgs.txt .
+RUN vcs import src < navigation2.repos
```
Comment (asimonov) [inline navigation2/Dockerfile:51]: one side effect from mentioning navigation2-release multiple times in repos file is that after vcs import we have a 35*108mb=~3.7gb of git commit history in src.

we do not need it for build

but i think the way docker works is unless we use external cache (like for apt) or remove it within the same RUN command, then it is resulting in permanent increase of image size.

so I propose we do this here:

```
RUN vcs import src < navigation2.repos && \
    sudo find . -name .git -print0 | xargs -0 rm -rf
```

(unless there is an option for vcs import to not bring .git in the first place)

Comment (xfiderek) [reply navigation2/Dockerfile:51]: good point, I think `vs import --shallow` should do the job. testing it rn

**Comment by xfiderek** [comment]
> May be more general question is in order here...
> 
> we are using rosinstall_generator (not just here, but in space-ros too), but the ros2 core has not fully switched to it yet, but uses ros2.repos file (https://docs.ros.org/en/jazzy/Installation/Alternatives/Ubuntu-Development-Setup.html#build-ros-2).
> 
> The issue was discussed here: [ros2/ros2_documentation#1470](https://github.com/ros2/ros2_documentation/issues/1470), but it was de-scoped from humble.
> 
> Does anyone know the status of this? rosinstall_generator obviously works for ros2, but it is still not official way...

We are using `rosinstall_generator` "offline", to generate a ros2.repos (navigation2.repos) file, so in a sense we are using .repos file as well to pull dependencies. rosinstall_generator is only used as a convenient tool to generate .repos file.

**Comment by asimonov** [comment]
> Hi @mkhansenbot , @Bckempa, @asimonov, other potential reviewers
> 
> I tried dividing it as we initially anticipated, but due to the problem with apt/from-source installation (dependency on nav2_bringup) it is really hard to make pinned repos working accurately without introducing space_nav2_bringup package. Is it ok if I introduce both pinned repos and space_nav2_bringup as part of this PR? Otherwise I would need to rework this significantly, only to rollback to what we have here. I will remove rviz2 changes from PR to keep reworks to the demo minimal.
> 
> I tested it locally and it is ready for review, but I think we need to wait for #162 before merging it, as we can't test it with the latest version of space robots image from main now due to using $ROSDISTRO instead of $ROS_DISTRO there.
> 
> So to sum it up:
> 
> * this PR is ready for review - it introduces space_nav2_bringup and pins repos, closing [Need to pin the Navigation2 version number for stability #142](https://github.com/space-ros/docker/issues/142)
> * my next PR (which I will create after this one gets merged) will remove dep on rviz2 and close [Update nav2_demo to match the new build approach demos#28](https://github.com/space-ros/demos/issues/28)
> * one more PR and corresponding issue will be created to remove gazebo from space_robots
> 
> Remember that you need to build the latest spaceros image locally for it to work

I agree about nav2_bringup. I am ok to keep space_nav2_bringup in this PR.

I will review later.

File: navigation2/Dockerfile (hunk 3 above)
Comment (asimonov) [reply navigation2/Dockerfile]: i did not know about 'shallow' option. did you find it in source code? there seem to be not much docs for vcstool

Comment (xfiderek) [reply navigation2/Dockerfile]: Frankly I dont remember because I found it a while ago :D. It should be listed with `vcs import --help` though

**Comment by asimonov** [comment]
I was about to verify this PR works by rebuilding almost all images we have from other in-flight PRs, but alas the git-lfs change interfered:

https://github.com/space-ros/simulation/issues/35

**Comment by xfiderek** [comment]
I rebased, resolved conflicts, and pushe changes. testing it locally.

@EzraBrooks I can see that commit message standard action failed. I believe its because I opened pull request from my own repo. All the PRs that were merged after adding this check were opened from the same repo, and I believe that is why it worked. I think this CI check should be fixed, wdyt?

File: navigation2/src/space_nav2_bringup/launch/localization_launch.py
Comment (EzraBrooks) [inline navigation2/src/space_nav2_bringup/launch/localization_launch.py:1]: Would this be simpler as an XML launch file?

**Comment by asimonov** [comment]
> I rebased, resolved conflicts, and pushed changes. testing it locally.
> 
> UPDATE: I re-tested changes and nav2 demo works fine. the only problem is that I had to setup `export MESA_LOADER_DRIVER_OVERRIDE=softpipe` for every GUI app in the docker container. I am not sure if its local issue, so I'd appreciate if you could run it on your pc @asimonov. Anyway, this problem is not related to that PR as I am facing it since the first time I run space robots demo locally
> 
> BTW, @EzraBrooks I can see that the commit message CI checker failed. I believe its because I opened pull request from my own repo. I think this CI check should be fixed, wdyt? All the PRs that were merged after adding this check were opened from `space-ros/docker` repo, and I believe that is why CI was green for them.

I can confirm I can run the demo.

But for some reason in the last step with rviz it shows there is no transform for 'map'.
I had to clone navigation2 repo into docker and use the default rviz config from navigation2 because the rviz config did not make it into the changed docker. Not sure if it was different from default, but it is not docker as we do not bring nav2_bringup there. So may be put it into space_nav2_bringup/config?

**Comment by asimonov** [comment]
> > I rebased, resolved conflicts, and pushed changes. testing it locally.
> > UPDATE: I re-tested changes and nav2 demo works fine. the only problem is that I had to setup `export MESA_LOADER_DRIVER_OVERRIDE=softpipe` for every GUI app in the docker container. I am not sure if its local issue, so I'd appreciate if you could run it on your pc @asimonov. Anyway, this problem is not related to that PR as I am facing it since the first time I run space robots demo locally
> > BTW, @EzraBrooks I can see that the commit message CI checker failed. I believe its because I opened pull request from my own repo. I think this CI check should be fixed, wdyt? All the PRs that were merged after adding this check were opened from `space-ros/docker` repo, and I believe that is why CI was green for them.
> 
> I can confirm I can run the demo.
> 
> But for some reason in the last step with rviz it shows there is no transform for 'map'. I had to clone navigation2 repo into docker and use the default rviz config from navigation2 because the rviz config did not make it into the changed docker. Not sure if it was different from default, but it is not docker as we do not bring nav2_bringup there. So may be put it into space_nav2_bringup/config?
> 
> i am talking about this step:
> 
> https://github.com/space-ros/docker/blob/5c8467d82b7f89aa072c14ed77166219135684a0/nav2_demo/README.md?plain=1#L44

I did another attepmt to run navigation demo.
Basically I can start navigation and localization launch files.
And I use the rviz config from humble branch of nav2 (https://raw.githubusercontent.com/ros-navigation/navigation2/refs/heads/humble/nav2_bringup/rviz/nav2_default_view.rviz) because we dont have rviz config migrated in this PR.
I can see costmaps appear but when i send a goal I see the following error message in rviz2 window:

```
[INFO] [1728656495.856096390] [rviz_navigation_dialog_action_client]: NavigateToPose will be called using the BT Navigator's default behavior tree.
2024-10-11 14:21:35.856 [RTPS_READER_HISTORY Error] Change payload size of '32' bytes is larger than the history payload size of '19' bytes and cannot be resized. -> Function can_change_be_added_nts
[ERROR] [1728656495.956606773] [rviz_navigation_dialog_action_client]: Send goal call failed
```

did anyone run navigation2 demo successfully? I was never able to do it...

File: navigation2/src/space_nav2_bringup/launch/localization_launch.py
Comment (asimonov) [reply navigation2/src/space_nav2_bringup/launch/localization_launch.py:1]: possibly, but we can do it later outside of this PR. I think launch file works as it is. we can replace as part of future migration to Gazebo Harmonic.

**Review by asimonov — APPROVED** [review]
I think it works. But would be nice to have rviz config included from nav2 into space_nav2_bringup, otherwise it is missing. We still have rviz2 from apt in the image, so the instructions are actually valid.

File: navigation2/Dockerfile [hunk 4]
```diff
@@ -41,69 +41,36 @@ LABEL org.label-schema.vcs-ref=${VCS_REF}
...
+    source ${SPACEROS_DIR}/install/setup.bash && \
+    sudo apt update && \
+    rosdep install -i --from-path src --skip-keys $(cat excluded-pkgs.txt) -y
+RUN source ${SPACEROS_DIR}/install/setup.bash && \
+    colcon build --cmake-args -DCMAKE_BUILD_TYPE=Release -DCMAKE_EXPORT_COMPILE_COMMANDS=ON --no-warn-unused-cli
+RUN rm -rf src build log
+
+
+# Install rviz2 to send goals to Nav2. 
+# TODO(xfiderek): Remove rviz2 and humble-nav2 froms this image
```
Comment (asimonov) [inline navigation2/Dockerfile:69]: I think as part of this we'll have to move nav2_default_view.rviz into space_nav2_bringup package.
Shall we do it as part of this PR?

**Comment by asimonov** [comment]
> I think it works. But would be nice to have rviz config included from nav2 into space_nav2_bringup, otherwise it is missing. We still have rviz2 from apt in the image, so the instructions are actually valid.

I tried the navigation demo again and it worked... i would say i probaby did something wrong last time trying to juggle 2 containers from different branches, as well as few terminals.

one observation: `nav2_default_view.rviz` at the moment comes from /opt/ros/humble - there is nav2_bringup package there. it comes there via apt install of rviz2 and nav2-bringup... I see it is still marked as TODO in nav2 Dockerfile. 
may be we need to finish it as part of this PR. at least copy rviz config so it is more self-contained. even if we dont add rviz2 to nav2 container... or may be we can just change instructions to pull vanilla nav2 docker and launch vanilla nav2 bringup rviz?

File: navigation2/Dockerfile (hunk 4 above)
Comment (xfiderek) [reply navigation2/Dockerfile:69]: Hi @asimonov, Following the comment of @mkhansenbot I suggested to remove rviz2 in a separate PR, to limit the changes, as this PR is already v. large. You are 100% right that it should be in that package, but I would suggest to do it as a separate PR as demo should work with current instructions and I would love to merge it soon, because its been open for quite some time. Removing rviz and including rviz configs into space_nav2_bringup could be tackled as a separate issue. are you ok with that?

**Comment by xfiderek** [comment]
> > I think it works. But would be nice to have rviz config included from nav2 into space_nav2_bringup, otherwise it is missing. We still have rviz2 from apt in the image, so the instructions are actually valid.
> 
> I tried the navigation demo again and it worked... i would say i probaby did something wrong last time trying to juggle 2 containers from different branches, as well as few terminals.
> 
> one observation: `nav2_default_view.rviz` at the moment comes from /opt/ros/humble - there is nav2_bringup package there. it comes there via apt install of rviz2 and nav2-bringup... I see it is still marked as TODO in nav2 Dockerfile. may be we need to finish it as part of this PR. at least copy rviz config so it is more self-contained. even if we dont add rviz2 to nav2 container... or may be we can just change instructions to pull vanilla nav2 docker and launch vanilla nav2 bringup rviz?

Hi @asimonov, good to know that it works!

Following the comment of @mkhansenbot I suggested to remove rviz2 in a separate PR, to limit the changes, as this PR is already v. large. You are 100% right that it should be changed, but I would suggest to do it as a separate PR as demo should work with current instructions and I would love to merge it soon, because its been open for quite some time. 
Also I am not sure if these configs would be part of space_nav2_bringup, as rviz would probably be launched from separate docker container. are you ok with that?

**Comment by xfiderek** [comment]
@EzraBrooks any update on commitee message standards CI checker? It seems that it is the only blocker now.
Maybe for now you could manually review commits and disable this rule for this PR if  we don't have time to fix the CI.

**Comment by asimonov** [comment]
> > > I think it works. But would be nice to have rviz config included from nav2 into space_nav2_bringup, otherwise it is missing. We still have rviz2 from apt in the image, so the instructions are actually valid.
> > 
> > 
> > I tried the navigation demo again and it worked... i would say i probaby did something wrong last time trying to juggle 2 containers from different branches, as well as few terminals.
> > one observation: `nav2_default_view.rviz` at the moment comes from /opt/ros/humble - there is nav2_bringup package there. it comes there via apt install of rviz2 and nav2-bringup... I see it is still marked as TODO in nav2 Dockerfile. may be we need to finish it as part of this PR. at least copy rviz config so it is more self-contained. even if we dont add rviz2 to nav2 container... or may be we can just change instructions to pull vanilla nav2 docker and launch vanilla nav2 bringup rviz?
> 
> Hi @asimonov, good to know that it works!
> 
> Following the comment of @mkhansenbot I suggested to handle rviz2 in a separate PR to limit the changes, as this PR is already v. large. You are 100% right that it should be changed, but I would suggest to do it as a separate PR as demo should work with current instructions and I would love to merge it soon, because its been open for quite some time. Also I am not sure if these configs would be part of space_nav2_bringup, as rviz would probably be launched from separate docker container. are you ok with that?
> 
> And just to confirm - you can run a demo without additional changes (e.g. cloning nav2 repos etc), right? You should be able run the demo as-is, set 2d pose estimate via rviz and then send nav2 goal (which should be followed by rover in gz).

yes, correct, i was able to run just by following the modified README, no other changes

ok, i am happy to merge in current state and then focus on cleaning up rviz dependency in separate issue

**Comment by eholum** [comment]
> @EzraBrooks any update on commitee message standards CI checker?

Fixed! Is this good to go then?

File: navigation2/Dockerfile (hunk 4 above)
Comment (xfiderek) [reply navigation2/Dockerfile:69]: As explained below, it will be tackled as a separate issue

File: navigation2/src/space_nav2_bringup/launch/localization_launch.py
Comment (xfiderek) [reply navigation2/src/space_nav2_bringup/launch/localization_launch.py:1]: Lets stick with python for now, as @asimonov agrees that it is fine. we can consider moving to XML as a separate disucssion

**Comment by xfiderek** [comment]
> > @EzraBrooks any update on commitee message standards CI checker?
> 
> Fixed! Is this good to go then?

@eholum I resolved all conversations. @asimonov tested it locally and it works for him. Lets merge it!

**Review by eholum — APPROVED** [review]
